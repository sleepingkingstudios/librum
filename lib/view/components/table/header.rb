# frozen_string_literal: true

module View::Components
  # Component for rendering a table header.
  class Table::Header < ViewComponent::Base
    # @param columns [Array<View::Components::DataField::FieldDefinition>] the
    #   columns used to render the table.
    def initialize(columns:, **)
      super()

      @columns = columns.map do |column|
        if column.is_a?(View::Components::DataField::FieldDefinition)
          column
        else
          View::Components::DataField::FieldDefinition.new(**column)
        end
      end
    end

    # @return [Array<View::Components::DataField::FieldDefinition>] the columns
    #   used to render the table.
    attr_reader :columns
  end
end
