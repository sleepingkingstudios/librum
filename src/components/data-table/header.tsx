import * as React from 'react';
import { kebabCase } from 'lodash';

import { joinClassNames } from '@utils/react-utils';
import { DataTableColumn } from './types';
import {
  alignmentClass,
  labelFor,
} from './utils';

const DataTableHeaderCell = ({
  column,
}: {
  column: DataTableColumn,
}): JSX.Element => {
  const joinedClassName = joinClassNames(
    `data-table-header-${kebabCase(column.name)}`,
    alignmentClass(column),
  );
  const label = labelFor({ column });

  return (
    <th className={joinedClassName}>{ label }</th>
  );
};

export const DataTableHeader = ({
  columns,
}: {
  columns: DataTableColumn[],
}): JSX.Element => {
  return (
    <thead className="data-table-header">
      <tr>
        {
          columns.map((column: DataTableColumn): JSX.Element => (
            <DataTableHeaderCell key={column.name} column={column} />
          ))
        }
      </tr>
    </thead>
  );
};
