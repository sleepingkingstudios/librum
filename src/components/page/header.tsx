import * as React from 'react';

import { FancyHr } from '@components/fancy-hr';

interface PageHeaderProps {
  title?: string;
  subtitle?: string;
}

const renderSubtitle = ({ subtitle }: { subtitle?: string; }): JSX.Element | null => {
  if (!(typeof subtitle === 'string')) { return null; }

  return (
    <React.Fragment>
      <br className="block sm:hidden" />

      <small className="m-0 sm:ml-2 text-2xl">{ subtitle }</small>
    </React.Fragment>
  )
}

export const PageHeader = ({
  title = 'Librum',
  subtitle = 'Campaign Companion'
}: PageHeaderProps): JSX.Element => {
  return (
    <header className="font-serif leading-9 mb-2 text-5xl text-center sm:text-left">
      { title }

      { renderSubtitle({ subtitle }) }

      <FancyHr className="mt-1" />
    </header>
  )
}
