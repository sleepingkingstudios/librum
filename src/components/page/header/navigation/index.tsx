import * as React from 'react';
import { map } from 'lodash';

import { joinClassNames } from '@utils/react-utils';
import {
  Dropdown,
  DropdownProps,
} from './dropdown';
import {
  NavigationItem,
  NavigationItemProps,
} from './item';
import { useThemeStyles } from '@themes';

export type NavigationProps = Array<NavigationItemProps | DropdownProps>;

interface PageNavigationProps {
  className?: string;
  navigation: NavigationProps;
}

const defaultListClassName = 'flex flex-col sm:flex-row';

export const PageNavigation = (
  { className, navigation }: PageNavigationProps
): JSX.Element => {
  const joinedClassNames = joinClassNames(
    'relative z-10',
    useThemeStyles('navigation'),
    className,
  );
  const joinedListNames = joinClassNames(
    defaultListClassName,
    useThemeStyles('navigation-list'),
  );
  const [open, setOpen] = React.useState<string | null>(null);

  return (
    <nav className={joinedClassNames}>
      <ul className={joinedListNames}>
        {
          map(
            navigation,
            (item: DropdownProps | NavigationItemProps): JSX.Element => {
              const { label } = item;

              if('items' in item) {
                return (
                  <Dropdown
                    key={label}
                    open={open}
                    setOpen={setOpen}
                    {...item}
                  />
                );
              }

              return (<NavigationItem key={label} {...item} />);
            },
          )
        }
      </ul>
    </nav>
  );
};
