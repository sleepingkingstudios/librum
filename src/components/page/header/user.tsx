import * as React from 'react';
import { FontAwesomeIcon as Icon } from '@fortawesome/react-fontawesome';
import { faUser } from '@fortawesome/free-solid-svg-icons';

import {
  actions,
  selector,
} from '@session';
import type { Session } from '@session';
import {
  useDispatch,
  useSelector,
} from '@store';
import { useThemeStyles } from '@themes';
import { joinClassNames } from '@utils/react-utils';

export const PageHeaderUser = (): JSX.Element => {
  const dispatch = useDispatch();
  const logout = () => { dispatch(actions.destroy()); };
  const session: Session = useSelector(selector);
  const {
    authenticated,
    user,
  } = session;

  const joinedClassNames = joinClassNames(
    'float-right',
    useThemeStyles('linkDanger'),
  );

  if (!authenticated) { return null; }

  const { username } = user;

  return (
    <div className="text-left">
      <Icon icon={faUser} className='mr-1' />

      <span className="hidden sm:inline">You are currently logged in as </span>

      { username }

      <span className={joinedClassNames}>
        <button onClick={logout}>Log Out</button>
      </span>
    </div>
  )
};
