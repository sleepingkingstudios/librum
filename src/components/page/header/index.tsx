import * as React from 'react';

import { PageNavigation } from '../navigation';
import type { NavigationProps } from '../navigation';
import { PageUser } from '../user';

interface IPageHeaderProps {
  title?: string;
  subtitle?: string;
  navigation?: NavigationProps | JSX.Element;
}

interface IRenderNavigation {
  navigation?: NavigationProps | JSX.Element;
}

const defaultClassName = "mb-3 text-center sm:text-left";

const PageHeaderSubtitle = ({ subtitle }: { subtitle?: string; }): JSX.Element | null => {
  if (!(typeof subtitle === 'string')) { return null; }

  return (
    <React.Fragment>
      <br className="block sm:hidden" />

      <small className="page-subtitle m-0 sm:ml-2">{ subtitle }</small>
    </React.Fragment>
  )
}

const renderNavigation = ({ navigation }: IRenderNavigation): JSX.Element => {
  if (navigation === null || navigation === undefined) { return null; }

  if (!('length' in navigation)) { return navigation; }

  if (navigation.length === 0) { return null; }

  return (
    <>
      <hr className="hr-fancy my-1 block sm:hidden" />

      <PageNavigation navigation={navigation} />
    </>
  );
};

export const PageHeader = ({
  title = 'Librum',
  subtitle = 'Campaign Companion',
  navigation
}: IPageHeaderProps): JSX.Element => {
  return (
    <header className={defaultClassName}>
      <span className="page-title">{ title }</span>

      <PageHeaderSubtitle subtitle={subtitle} />

      { renderNavigation({ navigation }) }

      <hr className="hr-fancy my-1" />

      <PageUser />
    </header>
  );
};
