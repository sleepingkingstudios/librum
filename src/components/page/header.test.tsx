import * as React from 'react';

import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { PageHeader } from './header';

describe('<PageHeader>', () => {
  it('should display the default title', () => {
    const defaultTitle = 'Librum';

    render(<PageHeader />);

    const text = screen.getByText(defaultTitle);

    expect(text).toBeVisible();
  });

  it('should display the default subtitle', () => {
    const defaultSubtitle = 'Campaign Companion';

    render(<PageHeader />);

    const text = screen.getByText(defaultSubtitle);

    expect(text).toBeVisible();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(<PageHeader />);

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with subtitle: null', () => {
    it('should not display the subtitle', () => {
      const defaultTitle = 'Librum';

      render(<PageHeader subtitle={null} />);

      const text = screen.getByText(defaultTitle);

      expect(text).toBeVisible();
      expect(text).toHaveTextContent(/^Librum$/);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(<PageHeader />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with subtitle: value', () => {
    const subtitle = 'Example Subtitle';

    it('should display the configured title', () => {
      render(<PageHeader subtitle={subtitle} />);

      const text = screen.getByText(subtitle);

      expect(text).toBeVisible();
    });
  });

  describe('with title: value', () => {
    const title = 'Example Title';

    it('should display the configured title', () => {
      render(<PageHeader title={title} />);

      const text = screen.getByText(title);

      expect(text).toBeVisible();
    });
  });
});
