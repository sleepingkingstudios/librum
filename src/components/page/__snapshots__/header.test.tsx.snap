// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageHeader> should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="font-serif leading-9 mb-2 text-5xl text-center sm:text-left"
  >
    Librum
    <br
      class="block sm:hidden"
    />
    <small
      class="m-0 sm:ml-2 text-2xl"
    >
      Campaign Companion
    </small>
    <hr
      class="mt-1 fancy_hr"
    />
  </header>
</DocumentFragment>
`;

exports[`<PageHeader> with subtitle: null should match the snapshot 1`] = `
<DocumentFragment>
  <header
    class="font-serif leading-9 mb-2 text-5xl text-center sm:text-left"
  >
    Librum
    <br
      class="block sm:hidden"
    />
    <small
      class="m-0 sm:ml-2 text-2xl"
    >
      Campaign Companion
    </small>
    <hr
      class="mt-1 fancy_hr"
    />
  </header>
</DocumentFragment>
`;
